packages/compiler/src/experimental/mutators.ts

```diff
@@ -0,0 +1,296 @@
+import { Program } from "../core/program.js";
+import {
+  Decorator,
+  Enum,
+  EnumMember,
+  FunctionParameter,
+  FunctionType,
+  Interface,
+  IntrinsicType,
+  Model,
+  ModelProperty,
+  Namespace,
+  ObjectType,
+  Operation,
+  Projection,
+  Scalar,
+  ScalarConstructor,
+  StringTemplate,
+  StringTemplateSpan,
+  TemplateParameter,
+  Tuple,
+  Type,
+  Union,
+  UnionVariant,
+} from "../core/types.js";
+import { CustomKeyMap } from "../emitter-framework/custom-key-map.js";
+import { Realm } from "./realm.js";
+import { $ } from "./typekit/index.js";
+
+/** @experimental */
+export type MutatorRecord<T extends Type> =
+  | {
+      filter?: MutatorFilterFn<T>;
+      mutate: MutatorFn<T>;
+    }
+  | {
+      filter?: MutatorFilterFn<T>;
+      replace: MutatorReplaceFn<T>;
+    }
+  | MutatorFn<T>;
+
+/** @experimental */
+export interface MutatorFn<T extends Type> {
+  (sourceType: T, clone: T, program: Program, realm: Realm): void;
+}
+
+/** @experimental */
+export interface MutatorFilterFn<T extends Type> {
+  (sourceType: T, program: Program, realm: Realm): boolean | MutatorFlow;
+}
+
+/** @experimental */
+export interface MutatorReplaceFn<T extends Type> {
+  (sourceType: T, clone: T, program: Program, realm: Realm): Type;
+}
+
+/** @experimental */
+export interface Mutator {
+  name: string;
+  Model?: MutatorRecord<Model>;
+  ModelProperty?: MutatorRecord<ModelProperty>;
+  Scalar?: MutatorRecord<Scalar>;
+  Enum?: MutatorRecord<Enum>;
+  EnumMember?: MutatorRecord<EnumMember>;
+  Union?: MutatorRecord<Union>;
+  UnionVariant?: MutatorRecord<UnionVariant>;
+  Tuple?: MutatorRecord<Tuple>;
+  Operation?: MutatorRecord<Operation>;
+  Interface?: MutatorRecord<Interface>;
+  String?: MutatorRecord<Scalar>;
+  Number?: MutatorRecord<Scalar>;
+  Boolean?: MutatorRecord<Scalar>;
+  ScalarConstructor?: MutatorRecord<ScalarConstructor>;
+  StringTemplate?: MutatorRecord<StringTemplate>;
+  StringTemplateSpan?: MutatorRecord<StringTemplateSpan>;
+}
+
+/** @experimental */
+export enum MutatorFlow {
+  MutateAndRecurse = 0,
+  DoNotMutate = 1 << 0,
+  DoNotRecurse = 1 << 1,
+}
+
+/** @experimental */
+export type MutableType = Exclude<
+  Type,
+  | TemplateParameter
+  | Namespace
+  | IntrinsicType
+  | FunctionType
+  | Decorator
+  | FunctionParameter
+  | ObjectType
+  | Projection
+>;
+const typeId = CustomKeyMap.objectKeyer();
+const mutatorId = CustomKeyMap.objectKeyer();
+const seen = new CustomKeyMap<[MutableType, Set<Mutator> | Mutator[]], Type>(([type, mutators]) => {
+  const key = `${typeId.getKey(type)}-${[...mutators.values()]
+    .map((v) => mutatorId.getKey(v))
+    .join("-")}`;
+  return key;
+});
+
+/** @experimental */
+export function mutateSubgraph<T extends MutableType>(
```

Comment:
what do you think about also prefixing the experimental apis with `unsafe_` like we did with `useMapState`

Reply:
I think that makes sense

Reply:
I did this in the exports
![image](https://github.com/user-attachments/assets/49e9c8bb-1692-4c93-a795-6f8e0268e28e)